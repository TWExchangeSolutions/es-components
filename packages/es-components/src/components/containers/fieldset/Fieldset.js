import React from 'react';
import PropTypes from 'prop-types';
import styled, { ThemeProvider, withTheme } from 'styled-components';
import viaTheme from 'es-components-via-theme';

const Legend = styled.legend`
  border: 0;
  border-bottom: 1px solid ${props => props.theme.colors.grayDark};
  color: ${props => props.theme.colors.grayDark};
  display: block;
  font-size: ${props => props.theme.sizes.baseFontSize * 1.5}px;
  margin: 0 0 25px 0;
  padding: 0;
  width: 100%;
`;

function renderLegend(content, legendClasses) {
  return content ? <Legend className={legendClasses}>{content}</Legend> : null;
}

const StyledFieldset = styled.fieldset`
  border: 0;
  margin: 0;
  padding: 0;
`;

function Fieldset({
  legendClasses,
  legendContent,
  children,
  extraContent,
  theme
}) {
  return (
    <ThemeProvider theme={theme}>
      <StyledFieldset>
        <div>
          {renderLegend(legendContent, legendClasses)}
          {extraContent}
        </div>
        {children}
      </StyledFieldset>
    </ThemeProvider>
  );
}

Fieldset.propTypes = {
  /** Determine whether or not to add a legend and what content to display  */
  legendContent: PropTypes.node,
  /** Additional classes to be applied to the legend element */
  legendClasses: PropTypes.string,
  children: PropTypes.node,
  /** Extra content that can be rendered after the Legend but before the radio buttons, allows
   * content to be put in that will not affect the accessability of the Legend/input relationship.
   */
  extraContent: PropTypes.node,
  /**
   * Theme object used by the ThemeProvider,
   * automatically passed by any parent component using a ThemeProvider
   */
  theme: PropTypes.object
};

Fieldset.defaultProps = {
  theme: viaTheme
};

export default withTheme(Fieldset);
