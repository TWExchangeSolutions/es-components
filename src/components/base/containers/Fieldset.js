import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import { colors, sizes } from '../../theme';

const Legend = styled.legend`
  border: 0;
  border-bottom: 1px solid ${colors.grayDark};
  color: ${colors.grayDark};
  display: block;
  font-size: ${sizes.baseFontSize * 0.75}px;
  font-weight: bold;
  margin: 0 0 10px 0;
  padding: 0;
  text-transform: uppercase;
  width: 100%;
`;

function renderLegend(text, additionalLegendClasses) {
  return text ?
    <Legend className={additionalLegendClasses}>{text}</Legend> :
    null;
}

const StyledFieldset = styled.fieldset`
  border: 0;
  margin: 0;
  padding: 0;
`;

function Fieldset({
  additionalLegendClasses,
  legendText,
  children
}) {
  return (
    <StyledFieldset>
      { renderLegend(legendText, additionalLegendClasses) }
      { children }
    </StyledFieldset>
  );
}

Fieldset.propTypes = {
  /** Determine whether or not to add a legend and what text to display  */
  legendText: PropTypes.string,
  /** Additional classes to be applied to the legend element */
  additionalLegendClasses: PropTypes.string,
  children: PropTypes.node
};

export default Fieldset;
