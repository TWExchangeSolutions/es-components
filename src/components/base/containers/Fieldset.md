Use a ``Fieldset`` component to group related fields.

### Fieldset example without a legend:
```
<Fieldset>
  <FormGroup>
    <Textbox labelText="First name:" />
  </FormGroup>
  <FormGroup>
    <Textbox labelText="Middle initial:" />
  </FormGroup>
  <FormGroup>
    <Textbox labelText="Last name:" />
  </FormGroup>
</Fieldset>
```

### Fieldset with a legend

```
<Fieldset legendText="Please enter your name">
  <FormGroup>
    <Textbox labelText="First name:" />
  </FormGroup>
  <FormGroup>
    <Textbox labelText="Middle initial:" />
  </FormGroup>
  <FormGroup>
    <Textbox labelText="Last name:" />
  </FormGroup>
</Fieldset>
```