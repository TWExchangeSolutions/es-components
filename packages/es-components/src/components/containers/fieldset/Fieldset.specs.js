/* eslint-env jest */

import React from 'react';
import { mountWithTheme, renderWithTheme } from '../../../testing';

import Fieldset from './Fieldset';

describe('Fieldset component', () => {
  it('adds additional classes to legend element when aditionalLegendClasses prop is supplied', () => {
    const instance = mountWithTheme(
      <Fieldset legendContent="I am legend" legendClasses="legendary" />
    );
    expect(instance.find('legend').hasClass('legendary')).toBe(true);
  });

  it('renders as expected', () => {
    const tree = renderWithTheme(
      <Fieldset>
        <div>Fieldset child</div>
      </Fieldset>
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders as expected with legendText', () => {
    const tree = renderWithTheme(
      <Fieldset legendContent="I am legend">
        <div>Fieldset child</div>
      </Fieldset>
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders as expected with extra content', () => {
    const tree = renderWithTheme(
      <Fieldset legendContent="I am legend" extraContent="Extra Content">
        <div>Fieldset child</div>
      </Fieldset>
    );

    expect(tree).toMatchSnapshot();
  });
});
