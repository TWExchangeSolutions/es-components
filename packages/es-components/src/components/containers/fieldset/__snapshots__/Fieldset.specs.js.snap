// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fieldset component renders as expected 1`] = `
<fieldset
  className="sc-bwzfXH kzYDuJ"
>
  <div />
  <div>
    Fieldset child
  </div>
</fieldset>
`;

exports[`Fieldset component renders as expected with extra content 1`] = `
<fieldset
  className="sc-bwzfXH kzYDuJ"
>
  <div>
    <legend
      className="sc-bdVaJa JbwBs"
    >
      I am legend
    </legend>
    Extra Content
  </div>
  <div>
    Fieldset child
  </div>
</fieldset>
`;

exports[`Fieldset component renders as expected with legendText 1`] = `
<fieldset
  className="sc-bwzfXH kzYDuJ"
>
  <div>
    <legend
      className="sc-bdVaJa JbwBs"
    >
      I am legend
    </legend>
  </div>
  <div>
    Fieldset child
  </div>
</fieldset>
`;
